Add E2E tests for backfill pause and stop controls

KyeongJooni opened the pull request:
Add UI E2E tests for backfill pause, resume, and cancel controls

This PR adds end-to-end tests to verify the backfill control functionality on the DAG detail page.

**What this tests:**
- Pausing a running backfill and verifying status changes to paused
- Resuming a paused backfill and verifying status returns to running
- Cancelling an active backfill and verifying status changes to cancelled
- Verifying that cancelled backfills cannot be resumed

**Implementation:**
- Created `BackfillPage` page object following existing POM pattern
- Added test suite `backfill-controls.spec.ts` with 3 test cases
- All tests follow the same structure as existing E2E tests (`dags-list.spec.ts`)

closes: #59594

- While syncing/rebasing with upstream main on the previous PR, the diff unintentionally included files I didn’t intend to modify, which in turn triggered CODEOWNERS review requests broadly. To reduce noise and reviewer burden, I closed the original PR and opened this one with a cleaned-up, focused set of changes.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

<!-- Please keep an empty line above the dashes. -->

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

KyeongJooni committed:
feat(ui): rebase and fix E2E tests for backfill controls

Rebased on upstream/main to resolve conflicts and fix test failures.

- fix: use page.waitForLoadState() instead of waitForPageLoad()
- feat: add status verification with getBackfillStatus()
- refactor: update to new createBackfill(dagName, options) API
- refactor: apply global auth pattern from PR #59692
- refactor: use dynamic dates instead of hardcoded values

Closes #59594

KyeongJooni committed:
fix(ui): test backfill controls on running backfills

Avoid waitForNoActiveBackfill() timeout by testing controls on
actively running backfills instead of waiting for completion.
Add AUTH_FILE import and align with backfill.spec.ts patterns.

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/backfill-controls.spec.ts:
```diff
@@ -0,0 +1,121 @@
+/*!
+ * Licensed to the Apache Software Foundation (ASF) under one
```
We should add these test in backfill.spec.ts

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/backfill-controls.spec.ts:
```diff
@@ -0,0 +1,121 @@
+    await expect(backfillPage.cancelButton).not.toBeVisible({ timeout: 10_000 });
+  });
+
+  test("should not be able to resume a cancelled backfill", async ({ page }) => {
```
This test looks like same as `should cancel an active backfill`

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/backfill-controls.spec.ts:
```diff
@@ -0,0 +1,121 @@
+  return date.toISOString().slice(0, 16);
+};
+
+test.describe("Backfill Controls", () => {
```
We can use beforeEach here instead of triggering backfill code in every test
```

test.describe("Backfill controls", () => {
  test.describe.configure({ mode: "serial" }); // Prevents parallel conflicts
  
  let backfillPage: BackfillPage;
  const controlFromDate = getPastDate(15);
  const controlToDate = getPastDate(14);
  
  test.beforeEach(async ({ page }) => {
    backfillPage = new BackfillPage(page);
    
    await backfillPage.createBackfill(testDagId, {
      fromDate: controlFromDate,
      reprocessBehavior: "All Runs",
      toDate: controlToDate,
    });
    
    await backfillPage.navigateToBackfillsTab(testDagId);
  });
```

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/backfill-controls.spec.ts:
```diff
@@ -0,0 +1,121 @@
+
+test.describe("Backfill Controls", () => {
+  const testDagId = testConfig.testDag.id;
+  const fromDate = getPastDate(2);
```
```
test.describe("Backfill controls", () => {
  test.describe.configure({ mode: "serial" });
  test.setTimeout(180_000);

  const testDagId = testConfig.testDag.id;
  const controlFromDate = getPastDate(15);
  const controlToDate = getPastDate(14);

  let backfillPage: BackfillPage;

  test.beforeEach(async ({ page }) => {
    backfillPage = new BackfillPage(page);

    await backfillPage.createBackfill(testDagId, {
      fromDate: controlFromDate,
      reprocessBehavior: "All Runs",
      toDate: controlToDate,
    });

    await backfillPage.navigateToBackfillsTab(testDagId);
  });

  test.afterEach(async () => {
    try {
      await backfillPage.clickCancelButton();
    } catch {
    }
  });

  test("verify pause and resume backfill", async () => {
    await backfillPage.clickPauseButton();
    expect(await backfillPage.isBackfillPaused()).toBe(true);

    await backfillPage.clickPauseButton();
    expect(await backfillPage.isBackfillPaused()).toBe(false);
  });

  test("verify cancel backfill", async () => {
    await backfillPage.clickCancelButton();
    await expect(backfillPage.pauseButton).not.toBeVisible();
    await expect(backfillPage.cancelButton).not.toBeVisible();
  });
});
```

You can update the above in backfill-spec.ts file, and all these tests should be sufficient

KyeongJooni replied:
@vatsrahul1001 Updated per your feedback:

- Moved tests to backfill.spec.ts (removed backfill-controls.spec.ts)
- Implemented beforeEach/afterEach pattern with serial mode
- Replaced waitForTimeout with explicit state waits
- Removed unused methods (isBackfillDateErrorVisible, waitForBackfillCompletion)
- Removed duplicate test

Regarding the partial selector `aria-label*="ancel"`: kept it to support potential localized button labels, but happy to make it more specific if needed.

Thank you for the detailed review.

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/pages/BackfillPage.ts:
```diff
@@ -80,6 +83,9 @@ export class BackfillPage extends BasePage {
     this.backfillsTable = page.locator("table");
     this.backfillDateError = page.locator('text="Start Date must be before the End Date"');
+    this.backfillBanner = page.locator('div:has-text("Backfill in progress")');
+    this.cancelButton = page.locator('button[aria-label*="ancel"]');
```
Do we need partial selectors here?

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/pages/BackfillPage.ts:
```diff
@@ -102,6 +108,18 @@ export class BackfillPage extends BasePage {
+  public async clickCancelButton(): Promise<void> {
+    await expect(this.cancelButton).toBeVisible({ timeout: 10_000 });
+    await this.cancelButton.click();
+    await this.page.waitForTimeout(1000);
```
Replace `waitForTimeout(1000)` with explicit state waits like `await expect(button).not.toBeVisible()`

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/pages/BackfillPage.ts:
```diff
@@ -340,6 +369,10 @@ export class BackfillPage extends BasePage {
     await menuItem.click();
   }
 
+  public async waitForBackfillCompletion(timeout: number = 30_000): Promise<void> {
```
You can rebase and remove this method. We already have `waitForNoActiveBackfill` in main

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/pages/BackfillPage.ts:
```diff
@@ -290,6 +308,17 @@ export class BackfillPage extends BasePage {
     return await headers.count();
   }
 
+  public async isBackfillDateErrorVisible(): Promise<boolean> {
```
Lets remove this its not being used in tests. Also if needed we can use playwright isVisible() method

KyeongJooni committed:
refactor(ui): restructure backfill controls tests per review feedback

- Use beforeEach/afterEach pattern with serial mode
- Replace waitForTimeout with explicit state waits
- Remove unused helper methods
- Update test structure and naming conventions

KyeongJooni committed:
test(ui): use exact aria-label selectors for backfill controls

KyeongJooni committed:
fix(ui): wait for aria-label change after pause button click

KyeongJooni committed:
fix(ui): cancel existing backfill before creating new one in beforeEach

choo121600 commented on airflow-core/src/airflow/ui/tests/e2e/pages/BackfillPage.ts:
```diff
@@ -80,6 +83,9 @@ export class BackfillPage extends BasePage {
     this.backfillDateError = page.locator('text="Start Date must be before the End Date"');
+    this.backfillBanner = page.locator('div:has-text("Backfill in progress")');
+    this.cancelButton = page.locator('button[aria-label="Cancel backfill"]');
+    this.pauseButton = page.locator('button[aria-label="Pause backfill"], button[aria-label="Unpause backfill"]');
```
It looks like the static checks are failing.
You can find the logs here.
https://github.com/apache/airflow/actions/runs/20837962435/job/59874625961

KyeongJooni replied:
Will rebase on latest main to fix the i18n translation gaps!

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/backfill.spec.ts:
```diff
@@ -194,3 +194,55 @@ test.describe("validate date range", () => {
+  test.beforeEach(async ({ page }) => {
+    backfillPage = new BackfillPage(page);
+
+    // Cancel any existing backfill before creating a new one
+    await backfillPage.navigateToDagDetail(testDagId);
+    try {
+      await backfillPage.clickCancelButton();
+    } catch {
+      // No active backfill to cancel
+    }
```
Instead of empty try catch should we check visibility before clicking.
```suggestion
test.beforeEach(async ({ page }) => {
  backfillPage = new BackfillPage(page);
  await backfillPage.navigateToDagDetail(testDagId);
  
  // Cancel any existing backfill if visible
  if (await backfillPage.cancelButton.isVisible()) {
    await backfillPage.clickCancelButton();
  }
```

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/backfill.spec.ts:
```diff
@@ -194,3 +194,55 @@ test.describe("validate date range", () => {
+    await backfillPage.navigateToBackfillsTab(testDagId);
+  });
+
+  test.afterEach(async () => {
+    try {
+      await backfillPage.clickCancelButton();
+    } catch {
+    }
+  });
```
```suggestion
test.afterEach(async () => {
  // Cleanup: cancel backfill if still active
  if (await backfillPage.cancelButton.isVisible()) {
    await backfillPage.clickCancelButton();
  }
});
```

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/pages/BackfillPage.ts:
```diff
@@ -80,6 +83,9 @@ export class BackfillPage extends BasePage {
     this.backfillRunButton = page.locator('button:has-text("Run Backfill")');
     this.backfillsTable = page.locator("table");
     this.backfillDateError = page.locator('text="Start Date must be before the End Date"');
+    this.backfillBanner = page.locator('div:has-text("Backfill in progress")');
```
Looks like we are not using backfillBanner anywhere in tests

KyeongJooni replied:
Done! Removed the unused `backfillBanner` locator

KyeongJooni committed:
Update airflow-core/src/airflow/ui/tests/e2e/specs/backfill.spec.ts

Co-authored-by: Rahul Vats <43964496+vatsrahul1001@users.noreply.github.com>

KyeongJooni committed:
Update airflow-core/src/airflow/ui/tests/e2e/specs/backfill.spec.ts

Co-authored-by: Rahul Vats <43964496+vatsrahul1001@users.noreply.github.com>

KyeongJooni committed:
refactor(ui): remove unused backfillBanner locator

KyeongJooni committed:
Merge branch 'main' into e2e/backfill-controls

KyeongJooni committed:
fix(ui): make clickPauseButton more robust in E2E tests

- Increase visibility timeout from 10s to 30s for CI environments
- Add scrollIntoViewIfNeeded() to ensure button is in viewport
- Add explicit enabled state check before clicking
- Wait for network response to confirm pause action completed
- Follow DagsPage pattern for waitForResponse handling

KyeongJooni committed:
Merge branch 'main' into e2e/backfill-controls

choo121600 reviewed (changes requested):
There are still some static check failures related to linting/line length.
These should be easy to resolve by enabling prek locally😉

https://github.com/apache/airflow/blob/main/contributing-docs/08_static_code_checks.rst

KyeongJooni committed:
fix(ui): fix beforeAll to wait on DAG detail page and fix formatting

- Wait for backfill completion on DAG detail page where banner is visible
- Remove unnecessary findBackfillRowByDateRange call from beforeAll
- Fix indentation in beforeEach/afterEach blocks

KyeongJooni committed:
fix(ui): make waitForNoActiveBackfill more robust with page reload in E2E tests

KyeongJooni committed:
Merge branch 'main' into e2e/backfill-controls

choo121600 commented on airflow-core/src/airflow/ui/tests/e2e/pages/BackfillPage.ts:
```diff
@@ -80,6 +82,8 @@ export class BackfillPage extends BasePage {
     this.backfillsTable = page.locator("table");
     this.backfillDateError = page.locator('text="Start Date must be before the End Date"');
+    this.cancelButton = page.locator('button[aria-label="Cancel backfill"]');
+    this.pauseButton = page.locator('button[aria-label="Pause backfill"], button[aria-label="Unpause backfill"]');
```
```suggestion
    this.cancelButton = page.locator('button[aria-label="Cancel backfill"]');
    this.pauseButton = page.locator(
      'button[aria-label="Pause backfill"], button[aria-label="Unpause backfill"]',
    );
```

choo121600 replied:
It looks like the static check is failing at this point.

KyeongJooni committed:
Update airflow-core/src/airflow/ui/tests/e2e/pages/BackfillPage.ts

Co-authored-by: Yeonguk Choo <choo121600@gmail.com>

choo121600 commented on airflow-core/src/airflow/ui/tests/e2e/pages/BackfillPage.ts:
```diff
@@ -340,10 +386,20 @@ export class BackfillPage extends BasePage {
-    await expect(backfillInProgress).not.toBeVisible({ timeout: 300_000 });
+    await expect(async () => {
+      await page.reload();
+      await page.waitForLoadState("networkidle", { timeout: 30_000 });
```
Is `networkidle` necessary here?

As far as I know, `networkidle` is generally discouraged since it’s timing-based and can be flaky, so it might be worth being cautious about using it.

KyeongJooni replied:
Thanks for pointing this out. I've removed all networkidle usages and replaced them with explicit UI element waits.

choo121600 commented on airflow-core/src/airflow/ui/tests/e2e/specs/backfill.spec.ts:
```diff
@@ -56,12 +56,8 @@ test.describe("Backfill creation and validation", () => {
-        toDate: config.toDate,
-      });
-
+      // Wait for backfill to complete on DAG detail page (where the banner is visible)
```
```suggestion
      // Wait for backfill to complete on Dag detail page (where the banner is visible)
```

choo121600 replied:
To keep it consistent with other parts, it would be better to standardize DAG to Dag.

KyeongJooni committed:
fix(ui): replace waitForLoadState with element visibility assertions in E2E tests

KyeongJooni committed:
Update airflow-core/src/airflow/ui/tests/e2e/specs/backfill.spec.ts

Co-authored-by: Yeonguk Choo <choo121600@gmail.com>

KyeongJooni committed:
test(ui): add E2E test for cancelled backfill cannot be resumed

KyeongJooni committed:
fix(ui): wait for API response before checking modal close in E2E tests

KyeongJooni committed:
fix(ui): exclude dry_run and check success status in createBackfill E2E

KyeongJooni committed:
fix(ui): remove catch and increase timeout in createBackfill E2E

KyeongJooni committed:
fix(ui): improve error handling in createBackfill E2E test

KyeongJooni committed:
fix(ui): handle 409 conflict in createBackfill E2E by retrying

KyeongJooni committed:
fix(ui): wait for backdrop to disappear before clicking cancel button

vatsrahul1001 committed:
Merge branch 'main' into e2e/backfill-controls

KyeongJooni committed:
fix: force close modal if not closed after successful API response

KyeongJooni committed:
fix(ui): improve backfill e2e test stability

KyeongJooni committed:
fix(ui): extend backfill date range and add retry logic for pauseButton

choo121600 reviewed (approved):
Cool, Looks good!
All tests are passing, and this part looks fine to me.
Thanks for the hard work 🙂
